packages/nextjs/src/config/loaders/ast.ts

```diff
@@ -320,3 +328,169 @@ export function removeComments(ast: AST): void {
+
+/**
+ * Extracts all identifiers from an ArrayPattern within a destructured named export declaration
+ * statement ("export const [val] = [1]").
+ *
+ * This function recursively calls itself and `getExportIdentifiersFromObjectPattern` since destructuring assignments
+ * can be deeply nested with objects and arrays.
+ *
+ * Example:
+ * export const [{ foo: name1 }, [{ bar: [name2]}, name3]] = [{ foo: 1 }, [{ bar: [2] }, 3]];
+ *   --> ["name1", "name2", "name3"]
+ */
+function getExportIdentifiersFromArrayPattern(arrayPattern: jscsTypes.ArrayPattern): string[] {
+  const identifiers: string[] = [];
+
+  arrayPattern.elements.forEach(element => {
+    if (Identifier.check(element)) {
+      identifiers.push(element.name);
+    } else if (ObjectPattern.check(element)) {
+      identifiers.push(...getExportIdentifiersFromObjectPattern(element));
+    } else if (ArrayPattern.check(element)) {
+      identifiers.push(...getExportIdentifiersFromArrayPattern(element));
+    } else if (RestElement.check(element)) {
+      identifiers.push(...getExportIdentifiersFromRestElement(element));
+    }
+  });
+
+  return identifiers;
+}
+
+/**
+ * Grabs all identifiers from an ObjectPattern within a destructured named export declaration
+ * statement ("export const { val: name } = { val: 1 }").
+ *
+ * This function recursively calls itself and `getExportIdentifiersFromArrayPattern` since destructuring assignments
+ * can be deeply nested with objects and arrays.
+ *
+ * Example:
+ * export const { foo: [name1], bar: { baz: [{ quux: name2 }], ...name3 }} = { foo: [1], bar: { baz: [{ quux: 3 }]} };
+ *   --> ["name1", "name2", "name3"]
+ */
+function getExportIdentifiersFromObjectPattern(objectPatternNode: jscsTypes.ObjectPattern): string[] {
+  const identifiers: string[] = [];
+
+  objectPatternNode.properties.forEach(property => {
+    if (Property.check(property)) {
+      if (Identifier.check(property.value)) {
+        identifiers.push(property.value.name);
+      } else if (ObjectPattern.check(property.value)) {
+        identifiers.push(...getExportIdentifiersFromObjectPattern(property.value));
+      } else if (ArrayPattern.check(property.value)) {
+        identifiers.push(...getExportIdentifiersFromArrayPattern(property.value));
+      } else if (RestElement.check(property.value)) {
+        identifiers.push(...getExportIdentifiersFromRestElement(property.value));
+      }
+    } else if (RestElement.check(property)) {
+      identifiers.push(...getExportIdentifiersFromRestElement(property));
+    }
+  });
+
+  return identifiers;
+}
+
+/**
+ * Given the AST of a file, this function extracts all named exports from the file.
+ *
+ * @returns a list of deduplicated identifiers.
+ */
+export function getExportIdentifiers(ast: AST): string[] {
+  const identifiers: string[] = [];
+
+  const namedExportDeclarationNodes = ast
+    .find(ExportNamedDeclaration)
+    .nodes()
+    .map(namedExportDeclarationNode => namedExportDeclarationNode.declaration);
```

Comment:
```suggestion
  // or `exportNamedDeclarationDeclarationNodes`
  const exportNamedDeclarationDeclarations = ast
    .find(ExportNamedDeclaration)
    .nodes()
    .map(namedExportDeclarationNode => namedExportDeclarationNode.declaration);
```

It's clunky, but having a name which is close to but not totally matching the type led me, at least, to expect this to be what you'd get without the `.map()` call. (Or you could go the other way and call it `namedExports` or `namedExportNodes`.)

Reply:
Yup agree. I went for `namedExportDeclarationNodeDeclarations` since anything starting with `export` sounds like a verb which I didn't like. 41ebd28084dc7c79461541d04bc994d0c30fc264